cpp/src/sampling/uniform_neighbor_sampling_impl.hpp

```diff
@@ -115,13 +115,165 @@ uniform_nbr_sample_impl(
+    edge_weight_view ? std::make_optional(rmm::device_uvector<weight_t>(0, handle.get_stream()))
+                     : std::nullopt;
+  auto d_result_edge_type =
+    edge_type_view ? std::make_optional(rmm::device_uvector<edge_type_t>(0, handle.get_stream()))
+                   : std::nullopt;
+  rmm::device_uvector<int32_t> d_result_hop(0, handle.get_stream());
+  auto d_result_label = d_labels
+                          ? std::make_optional(rmm::device_uvector<int32_t>(0, handle.get_stream()))
+                          : std::nullopt;
+
+  int32_t hop{0};
+  size_t comm_size{1};
+
+  if constexpr (multi_gpu) {
+    seed += handle.get_comms().get_rank();
+    comm_size = handle.get_comms().get_size();
+  }
+
+  for (auto&& k_level : h_fan_out) {
+    // prep step for extracting out-degs(sources):
+    if constexpr (multi_gpu) {
+      if (d_labels) {
+        std::tie(d_in, *d_labels) = shuffle_int_vertices_and_values_by_gpu_id(
+          handle, std::move(d_in), std::move(*d_labels), graph_view.vertex_partition_range_lasts());
+      } else {
+        d_in = shuffle_int_vertices_by_gpu_id(
+          handle, std::move(d_in), graph_view.vertex_partition_range_lasts());
+      }
+    }
+
+    rmm::device_uvector<vertex_t> d_out_src(0, handle.get_stream());
+    rmm::device_uvector<vertex_t> d_out_dst(0, handle.get_stream());
+    rmm::device_uvector<edge_t> d_out_edge_id(0, handle.get_stream());
+    std::optional<rmm::device_uvector<weight_t>> d_out_weight{std::nullopt};
+    std::optional<rmm::device_uvector<edge_type_t>> d_out_edge_type{std::nullopt};
+    std::optional<rmm::device_uvector<int32_t>> d_out_label{std::nullopt};
+
+    if (k_level > 0) {
+      raft::random::RngState rng_state(seed);
+      seed += d_in.size() * k_level * comm_size;
+
+      std::tie(d_out_src, d_out_dst, d_out_edge_id, d_out_weight, d_out_edge_type, d_out_label) =
+        sample_edges(handle,
+                     graph_view,
+                     edge_weight_view,
+                     edge_type_view,
+                     rng_state,
+                     d_in,
+                     d_labels,
+                     static_cast<size_t>(k_level),
+                     with_replacement);
+    } else {
+      std::tie(d_out_src, d_out_dst, d_out_edge_id, d_out_weight, d_out_edge_type, d_out_label) =
+        gather_one_hop_edgelist(
+          handle, graph_view, edge_weight_view, edge_type_view, d_in, d_labels);
+    }
+
+    // resize accumulators:
+    auto old_sz = d_result_dst.size();
+    auto add_sz = d_out_dst.size();
+    auto new_sz = old_sz + add_sz;
+
+    d_result_src.resize(new_sz, handle.get_stream());
+    d_result_dst.resize(new_sz, handle.get_stream());
+    d_result_edge_id.resize(new_sz, handle.get_stream());
+    d_result_hop.resize(new_sz, handle.get_stream());
+    if (d_result_weight) d_result_weight->resize(new_sz, handle.get_stream());
+    if (d_result_edge_type) d_result_edge_type->resize(new_sz, handle.get_stream());
```

Comment:
Shouldn't we resize `d_result_label` as well?

i.e.
`if (d_result_label) d_result_label->resize(new_sz, handle.get_stream());`

Reply:
And note that up-sizing involves 1) reallocating a new array and 2) copying the old contents. Copying can add significant overhead if there are many levels.

We may better store the data for each level in a separate array and concatenate once at the end to avoid frequent resizing (and copying the same data multiple times).

Reply:
Added a FIXME to mention this performance issue.